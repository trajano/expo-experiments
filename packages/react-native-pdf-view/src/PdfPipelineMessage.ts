import { EventSubscription } from 'react-native';

export type OkPdfViewMessage = {
  type: 'ok';
  data: string;
  correlationId: string;
};
export type PageCountPdfViewMessage = {
  type: 'numPages';
  numPages: number;
  correlationId: string;
};
export type ViewPortPdfViewMessage = {
  type: 'viewport';
  width: number;
  height: number;
  scale: number;
  correlationId: string;
};
export type ErrorPdfViewMessage = {
  type: 'error';
  error: string;
  correlationId?: string;
  where?: string;
};

export type PdfPipelineMessage =
  | OkPdfViewMessage
  | PageCountPdfViewMessage
  | ViewPortPdfViewMessage
  | ErrorPdfViewMessage;

type PdfEventName = PdfPipelineMessage['type'];
type PdfPipelineListener = (event: PdfPipelineMessage) => void;
export type AddListenerFunction = (
  eventName: PdfEventName,
  correlationId: string,
  listener: PdfPipelineListener,
) => EventSubscription;
