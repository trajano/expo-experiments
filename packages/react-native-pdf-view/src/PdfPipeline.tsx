/**
 * A context that provides no specific functionality. This serves as a template for developers
 * to create React Context-based implementations and includes a higher-order component (HOC) version.
 */
import {
  ComponentType,
  createContext,
  FC,
  PropsWithChildren,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
} from 'react';
import { IpcWebProvider, useIpcWeb } from 'react-native-ipc-web';
import { renderPdfPipelineHtml } from './renderPdfPipelineHtml';
import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import EventEmitter from 'react-native/Libraries/vendor/emitter/EventEmitter';
import { EventSubscription } from 'react-native';
import { AddListenerFunction, PdfPipelineMessage } from './PdfPipelineMessage';

export interface PdfPipeline {
  /**
   * Adds an event listener
   * the event
   * @param eventName PDF EventName
   * @param correlationId correlation ID
   * @param listener event listener
   */
  addListener: AddListenerFunction;

  /**
   * @param correlationId correlation ID use Crypto.randomUUID()
   * @param pdfData base64 data for the PDF.
   * @param pageNumber defaults to 1
   * @param scale default to 1.0
   */
  postPdfRequest(
    correlationId: string,
    pdfData: string,
    pageNumber?: number,
    scale?: number,
  ): void;

  pdfPipelineReadyPromise: Promise<boolean>;
}

export const PdfPipelineContext = createContext<PdfPipeline>({
  addListener: () =>
    ({
      remove: () => {},
    }) as EventSubscription,
  postPdfRequest: () => {},
  pdfPipelineReadyPromise: Promise.resolve(true),
});

export const usePdfPipeline = () => useContext(PdfPipelineContext);

const InternalPdfPipelineProvider: FC<
  PropsWithChildren<{
    addListener: AddListenerFunction;
  }>
> = ({ addListener, children }) => {
  const { IpcWebView, postMessage, ipcWebViewReadyPromise } = useIpcWeb();

  const postPdfRequest = useCallback(
    (
      correlationId: string,
      pdfData: string,
      pageNumber: number = 1,
      scale: number = 1.0,
    ) => {
      postMessage({
        correlationId,
        pdfData,
        pageNumber,
        scale,
      });
    },
    [postMessage],
  );
  const value = useMemo<PdfPipeline>(
    () => ({
      addListener,
      postPdfRequest,
      pdfPipelineReadyPromise: ipcWebViewReadyPromise,
    }),
    [addListener, postPdfRequest, ipcWebViewReadyPromise],
  );
  return (
    <PdfPipelineContext.Provider value={value}>
      <IpcWebView />
      {children}
    </PdfPipelineContext.Provider>
  );
};
type PdfPipelineProviderProps = PropsWithChildren<{
  /**
   * Asset pointing to the pdf.js script, defaults to https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.8.69/pdf.min.mjs. It
   * can also be an asset reference.
   */
  pdfJs?: Asset;

  /**
   * Asset pointing to the pdf.js worker script, defaults to https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.8.69/pdf.worker.min.mjs.  It
   * can also be an asset reference.
   */
  pdfWorkerJs?: Asset;
}>;
export const PdfPipelineProvider: FC<PdfPipelineProviderProps> = ({
  pdfJs,
  pdfWorkerJs,
  children,
}) => {
  const eventEmitterRef = useRef(new EventEmitter());
  const renderSourceAsync = useCallback(async () => {
    const pdfJsAsset =
      pdfJs ??
      Asset.fromURI(
        'https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.8.69/pdf.min.mjs',
      );
    const pdfWorkerJsAsset =
      pdfWorkerJs ??
      Asset.fromURI(
        'https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.8.69/pdf.worker.min.mjs',
      );

    const [pdfJsCode, pdfWorkerBase64Code] = await Promise.all([
      FileSystem.readAsStringAsync(
        (await pdfJsAsset.downloadAsync()).localUri!,
        { encoding: 'utf8' },
      ),
      FileSystem.readAsStringAsync(
        (await pdfWorkerJsAsset.downloadAsync()).localUri!,
        { encoding: 'base64' },
      ),
    ]);
    return renderPdfPipelineHtml(
      pdfJsCode,
      `data:application/javascript;base64,${pdfWorkerBase64Code}`,
    );
  }, [pdfJs, pdfWorkerJs]);
  const onMessage = useCallback((message: PdfPipelineMessage) => {
    const eventEmitter = eventEmitterRef.current;
    // emit an event
    eventEmitter.emit(message.type, message);
  }, []);
  const addListener: AddListenerFunction = (
    eventName,
    correlationId,
    pdfPipelineEventListener,
  ) => {
    return eventEmitterRef.current.addListener(
      eventName,
      (event: PdfPipelineMessage) => {
        if (event.correlationId === correlationId) {
          pdfPipelineEventListener(event);
        }
      },
    );
  };
  useEffect(() => {
    const eventEmitter = eventEmitterRef.current;

    return () => {
      eventEmitter.removeAllListeners();
    };
  }, []);

  return (
    <IpcWebProvider sourceProvider={renderSourceAsync} onMessage={onMessage}>
      <InternalPdfPipelineProvider addListener={addListener}>
        {children}
      </InternalPdfPipelineProvider>
    </IpcWebProvider>
  );
};

export const WithPdfView = <P extends object>(
  Component: ComponentType<P>,
): FC<P> => {
  const WrappedComponent = ({ ...props }: P) => (
    <PdfPipelineProvider>
      <Component {...props} />
    </PdfPipelineProvider>
  );

  WrappedComponent.displayName = `WithPdfView(${Component.displayName ?? Component.name ?? 'Component'})`;

  return WrappedComponent;
};
