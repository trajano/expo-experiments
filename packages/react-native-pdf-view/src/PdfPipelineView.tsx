import { FC, useEffect, useState } from 'react';
import { Image } from 'expo-image';
import { fetchToBase64 } from './fetchToBase64';

import * as Crypto from 'expo-crypto';
import { usePdfPipeline } from './PdfPipeline';
import { PdfPipelineViewProps } from './PdfPipelineViewProps';

/**
 * This renders a PDF page using pdf.js inside a web view that's set up via PdfPipeline
 * The way it works is it generates a correlation ID, then sets up an event listener that is waiting for the PDF
 * event for the given correlation ID
 * Once it receives the image data it returns a ExpoImage
 */
export const PdfPipelineView: FC<PdfPipelineViewProps> = ({
  uri,
  pageNumber = 1,
  scale = 1.0,
  onViewPortKnown = () => {},
  onError = () => {},
  onPageCountKnown = () => {},
  contentFit = 'cover',
  style,
  ...props
}) => {
  const [imageDataUri, setImageDataUri] = useState<string | undefined>(
    undefined,
  );
  const {
    postPdfRequest,
    addListener: addPdfListener,
    pdfPipelineReadyPromise,
  } = usePdfPipeline();

  useEffect(() => {
    const correlationId = Crypto.randomUUID();
    const viewportListenerSubscription = addPdfListener(
      'viewport',
      correlationId,
      (event) => {
        if (event.type !== 'viewport') {
          return;
        }
        onViewPortKnown({
          width: event.width,
          height: event.height,
          scale: event.scale,
        });
        viewportListenerSubscription.remove();
      },
    );

    const pageCountListenerSubscription = addPdfListener(
      'numPages',
      correlationId,
      (event) => {
        if (event.type !== 'numPages') {
          return;
        }
        onPageCountKnown({
          pageCount: event.numPages,
        });
        pageCountListenerSubscription.remove();
      },
    );
    const okListenerSubscription = addPdfListener(
      'ok',
      correlationId,
      (event) => {
        if (event.type !== 'ok') {
          return;
        }
        // may cache this later
        setImageDataUri(event.data);
        okListenerSubscription.remove();
      },
    );
    const errorListenerSubscription = addPdfListener(
      'error',
      correlationId,
      (event) => {
        if (event.type !== 'error') {
          return;
        }
        onError({
          error: event.error,
        });
      },
    );
    (async () => {
      try {
        const data = await fetchToBase64(uri);
        await pdfPipelineReadyPromise;
        postPdfRequest(correlationId, data, pageNumber, scale);
      } catch (error: unknown) {
        console.error('posted', error);
      }
    })();
    return () => {
      okListenerSubscription.remove();
      errorListenerSubscription.remove();
      viewportListenerSubscription.remove();
    };
  }, [
    uri,
    addPdfListener,
    postPdfRequest,
    onPageCountKnown,
    onViewPortKnown,
    pageNumber,
    onError,
    pdfPipelineReadyPromise,
    scale,
  ]);
  return (
    <Image
      {...props}
      source={{ uri: imageDataUri }}
      style={style}
      contentFit={contentFit}
    />
  );
};
