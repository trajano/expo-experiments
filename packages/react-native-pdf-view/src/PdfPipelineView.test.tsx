import { render, act, screen } from '@testing-library/react-native';
import { PdfPipelineView } from './PdfPipelineView';
import { usePdfPipeline } from './PdfPipeline';
import * as Crypto from 'expo-crypto';
import { Asset } from 'expo-asset';
import { PdfPipelineMessage } from './PdfPipelineMessage';

// Mock dependencies
jest.mock('./PdfPipeline', () => ({
  usePdfPipeline: jest.fn(),
}));

jest.mock('expo-crypto', () => ({
  randomUUID: jest.fn(),
}));

jest.mock('expo-asset', () => ({
  Asset: {
    fromURI: jest.fn(),
  },
}));

describe('PdfPipelineView', () => {
  const mockPostPdfRequest = jest.fn();
  const mockAddListener = jest.fn();
  const mockPdfPipelineReadyPromise = Promise.resolve(true);

  beforeEach(() => {
    jest.clearAllMocks();
    (Crypto.randomUUID as jest.Mock).mockReturnValue('test-correlation-id');
    (usePdfPipeline as jest.Mock).mockReturnValue({
      postPdfRequest: mockPostPdfRequest,
      addListener: mockAddListener,
      pdfPipelineReadyPromise: mockPdfPipelineReadyPromise,
    });
    (Asset.fromURI as jest.Mock).mockReturnValue({
      downloadAsync: jest
        .fn()
        .mockResolvedValue({ localUri: 'file://test.pdf' }),
    });
  });

  it('renders without crashing', () => {
    const { getByTestId } = render(
      <PdfPipelineView
        uri="test-uri"
        pageNumber={1}
        scale={1.0}
        onViewPortKnown={jest.fn()}
        onError={jest.fn()}
        onPageCountKnown={jest.fn()}
        contentFit="cover"
        style={{ width: 100, height: 100 }}
        testID="pdf-pipeline-view"
      />,
    );

    expect(getByTestId('pdf-pipeline-view')).toBeTruthy();
  });

  it('sets up event listeners and posts PDF request', async () => {
    const { rerender } = render(
      <PdfPipelineView
        uri="test-uri"
        pageNumber={1}
        scale={1.0}
        onViewPortKnown={jest.fn()}
        onError={jest.fn()}
        onPageCountKnown={jest.fn()}
        testID="pdf-pipeline-view"
      />,
    );

    // Assert listeners and postPdfRequest were called
    await act(async () => {
      expect(mockAddListener).toHaveBeenCalledTimes(5);
    });
    await act(async () => {
      expect(mockPostPdfRequest).toHaveBeenCalledWith(
        'test-correlation-id',
        'file://test.pdf',
        1,
        1.0,
      );
    });

    expect(screen.getByTestId('pdf-pipeline-view')).toBeTruthy();
    expect(screen.getByTestId('pdf-pipeline-view').props.source).toEqual([]);

    const okCallback = mockAddListener.mock.calls.find(
      (it) => it[0] === 'ok',
    )[2];
    expect(typeof okCallback).toBe('function');

    await act(async () => {
      okCallback({
        type: 'ok',
        data: 'abc',
        correlationId: 'test-correlation-id',
      } satisfies PdfPipelineMessage);
    });
    expect(screen.getByTestId('pdf-pipeline-view')).toBeTruthy();
    expect(screen.getByTestId('pdf-pipeline-view').props.source).toContainEqual(
      { uri: 'abc' },
    );

    expect(screen.toJSON()).toMatchSnapshot();

    mockAddListener.mockClear();

    // Update props and ensure effect runs again
    rerender(
      <PdfPipelineView
        uri="new-test-uri"
        pageNumber={2}
        scale={1.5}
        onViewPortKnown={jest.fn()}
        onError={jest.fn()}
        onPageCountKnown={jest.fn()}
        testID="pdf-pipeline-view"
      />,
    );
    await act(async () => Promise.resolve());
    await act(async () => {
      expect(mockAddListener).toHaveBeenCalledTimes(5);
      expect(mockPostPdfRequest).toHaveBeenCalledWith(
        'test-correlation-id',
        'file://test.pdf',
        2,
        1.5,
      );
    });
    expect(screen.getByTestId('pdf-pipeline-view')).toBeTruthy();
    expect(screen.getByTestId('pdf-pipeline-view').props.source).toContainEqual(
      { uri: 'abc' },
    );

    const receivedCallback = mockAddListener.mock.calls.find(
      (it) => it[0] === 'received',
    )[2];
    expect(typeof okCallback).toBe('function');
    await act(async () => {
      receivedCallback({
        type: 'received',
        uri: 'new-test-uri',
        pageNumber: 2,
        scale: 1.5,
        correlationId: 'test-correlation-id',
      } satisfies PdfPipelineMessage);
    });
    expect(screen.getByTestId('pdf-pipeline-view').props.source).toEqual([]);

    const rerenderOkCallback = mockAddListener.mock.calls.find(
      (it) => it[0] === 'ok',
    )[2];

    await act(async () => {
      rerenderOkCallback({
        type: 'ok',
        data: 'new-abc',
        correlationId: 'test-correlation-id',
      } satisfies PdfPipelineMessage);
    });
    expect(screen.getByTestId('pdf-pipeline-view').props.source).toContainEqual(
      { uri: 'new-abc' },
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('cleans up listeners on unmount', () => {
    const mockRemoveListener = jest.fn();
    mockAddListener.mockReturnValue({ remove: mockRemoveListener });

    const { unmount } = render(
      <PdfPipelineView
        uri="test-uri"
        pageNumber={1}
        scale={1.0}
        onViewPortKnown={jest.fn()}
        onError={jest.fn()}
        onPageCountKnown={jest.fn()}
      />,
    );

    unmount();
    expect(mockRemoveListener).toHaveBeenCalledTimes(5);
  });
});
