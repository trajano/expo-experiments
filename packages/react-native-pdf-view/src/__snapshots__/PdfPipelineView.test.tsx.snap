// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PdfPipelineView sets up event listeners and posts PDF request 1`] = `
<ViewManagerAdapter_ExpoImage
  containerViewRef={"[React.ref]"}
  contentFit="cover"
  contentPosition={
    {
      "left": "50%",
      "top": "50%",
    }
  }
  nativeViewRef={"[React.ref]"}
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  onProgress={[Function]}
  placeholder={[]}
  source={
    [
      {
        "uri": "abc",
      },
    ]
  }
  style={{}}
  testID="pdf-pipeline-view"
  transition={null}
/>
`;

exports[`PdfPipelineView sets up event listeners and posts PDF request 2`] = `
<ViewManagerAdapter_ExpoImage
  containerViewRef={"[React.ref]"}
  contentFit="cover"
  contentPosition={
    {
      "left": "50%",
      "top": "50%",
    }
  }
  nativeViewRef={"[React.ref]"}
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  onProgress={[Function]}
  placeholder={[]}
  source={
    [
      {
        "uri": "new-abc",
      },
    ]
  }
  style={{}}
  testID="pdf-pipeline-view"
  transition={null}
/>
`;
